core/src/main/java/fr/sncf/osrd/speedcontroller/generators/ConstructionAllowanceGenerator.java

```diff
@@ -30,44 +33,45 @@ protected double getFirstLowEstimate() {
     }
 
     @Override
-    protected double getFirstHighEstimate() {
-        var speeds = getExpectedSpeeds(schedule, maxSpeedControllers, TIME_STEP);
+    protected double getFirstHighEstimate(SortedDoubleMap speeds) {
         return max(speeds.values()) * DICHOTOMY_MARGIN;
     }
 
     @Override
-    protected double getFirstGuess() {
-        // first guess is the mean value between max and min
-        return (this.getFirstHighEstimate() + this.getFirstLowEstimate()) / 2;
+    protected void initializeBinarySearch(TrainSchedule schedule, SortedDoubleMap speeds) {
+        physicalLimits = generatePhysicalLimits(schedule, speeds);
     }
 
     @Override
     protected Set<SpeedController> getSpeedControllers(TrainSchedule schedule,
-                                                       double targetSpeed,
-                                                       double initialPosition,
-                                                       double endPosition) throws SimulationError {
+                                                       SortedDoubleMap speeds,
+                                                       double targetSpeed) throws SimulationError {
+        return getSpeedControllers(schedule, speeds, targetSpeed, sectionBegin, sectionEnd);
+    }
 
-        // for later implementation, speed limit under which a train would consume too much capacity
-        final double capacitySpeedLimit = 30 / 3.6;
+    @Override
+    protected Set<SpeedController> getSpeedControllers(TrainSchedule schedule,
```

Comment:
This method is really long now, maybe we could try to split some parts of it into new private methods?